CefSharp/DevTools/DevToolsClient.cs

```diff
@@ -199,69 +273,85 @@ bool IDevToolsMessageObserver.OnDevToolsMessage(IBrowser browser, Stream message
+                    {
+                        context.SetResult(new DevToolsMethodResponse
+                        {
+                            Success = success,
+                            MessageId = messageId,
+                            ResponseAsJsonString = new StreamReader(result).ReadToEnd(),
+                        });
+                    }
+                    else
                     {
-                        taskCompletionSource.TrySetResult(methodResult);
-                    };
+                        try
+                        {
+                            context.SetResult(DeserializeJson(context.Type, result));
+                        }
+                        catch (Exception ex)
+                        {
+                            context.SetException(ex);
+                        }
+                    }
                 }
                 else
                 {
-                    execute = () =>
-                    {
-                        var errorObj = methodResult.DeserializeJson<DevToolsDomainErrorResponse>(ignoreSuccess: true);
-                        errorObj.MessageId = messageId;
+                    var errorObj = (DevToolsDomainErrorResponse)DeserializeJson(typeof(DevToolsDomainErrorResponse), result);
+                    errorObj.MessageId = messageId;
 
-                        //Make sure continuation runs on Thread Pool
-                        taskCompletionSource.TrySetException(new DevToolsClientException("DevTools Client Error :" + errorObj.Message, errorObj));
-                    };
+                    context.SetException(new DevToolsClientException("DevTools Client Error :" + errorObj.Message, errorObj));
                 }
+            }
+        }
+
+        private struct MethodResultContext
+        {
+            public readonly Type Type;
+            private readonly Func<object, bool> setResult;
+            private readonly Func<Exception, bool> setException;
+            private readonly SynchronizationContext syncContext;
+
+            public MethodResultContext(Type type, Func<object, bool> setResult, Func<Exception, bool> setException, SynchronizationContext syncContext)
+            {
+                Type = type;
+                this.setResult = setResult;
+                this.setException = setException;
+                this.syncContext = syncContext;
+            }
+
+            public void SetResult(object result)
+            {
+                InvokeOnSyncContext(setResult, result);
+            }
+
+            public void SetException(Exception ex)
+            {
+                InvokeOnSyncContext(setException, ex);
+            }
 
-                var syncContext = taskCompletionSource.SyncContext;
-                if (syncContext == null)
+            private void InvokeOnSyncContext<T>(Func<T, bool> fn, T value)
+            {
+                if (syncContext == null || syncContext == SynchronizationContext.Current)
```

Comment:
From memory SynchronizationContext.Current creates a new context if one doesn't exist. This is likely undesirable if called on a CEF thread.

Reply:
Extra check should be added to method execution to avoid creating sync context, I'll add that after this PR is complete.

Reply:
My mistake,  I'm thinking of the WPF dispatcher. This should be fine.